File: WORKSPACE/Assignment1/assignment1.R
#Author Aime Byiringiro

#We are  trying to predict if people in specific countries at certain age
#  will or will not purchase 
# This is a classification problem since the dependent variable is categorical. 


# Importing data from Customer_Data.csv in the working directory 
dataset = read.csv('Customer_Data.csv')

# replacing misssing Age with median of all ages
dataset$Age = ifelse(is.na(dataset$Age),
                     median(dataset$Age, na.rm = TRUE),
                     dataset$Age)
                     
#replacing missing salaires with mean of salaries 
dataset$Salary = ifelse(is.na(dataset$Salary),
                        mean(dataset$Salary, na.rm = TRUE),
                        dataset$Salary)
# 
# Encoding Country column data
#India = 1, Sri lanka = 2, China = 3
dataset$Country = factor(dataset$Country,
                         levels = c('India', 'Sri lanka', 'China'),
                         labels = c(1, 2, 3))

# Encoding Purchased column data 
#No = 0, Yes. =1
dataset$Purchased = factor(dataset$Purchased,
                           levels = c('No', 'Yes'),
                           labels = c(0, 1))

# 1/3 of the data will be used for test and ther rest for training
library(caTools)
set.seed(123) # for reproducibility
# we use the dependent variable to split becuase it is a classification problem
# if it was a regression problem we would use the independent variables

split = sample.split(dataset$Purchased, SplitRatio = 2/3)


training_set = subset(dataset, split == TRUE) # training set which is 2/3
test_set = subset(dataset, split == FALSE) # test set which is 1/3

# Feature Scaling
# Define min-max normalization function
min_max_norm <- function(x, min_val, max_val) {

  return((x - min_val) / (max_val - min_val))
}

# Compute min and max values for each feature in the training set

# explain why 2, 3, and 3 are used
# [, 2:3], 2, min means select columns 2 and 3 and apply min function to them 
# we use 2 and 3 because we want to normalize age and salary
# we dont want to normalize country and purchased because they are categorical

min_vals <- apply(training_set[, 2:3], 2, min)
max_vals <- apply(training_set[, 2:3], 2, max)

# Normalize training set using its own min-max values
training_set[, 2:3] <- as.data.frame(mapply(min_max_norm, 
                                            training_set[, 2:3], 
                                            min_vals, 
                                            max_vals))

# Normalize test set using training set's min-max values
test_set[, 2:3] <- as.data.frame(mapply(min_max_norm, 
                                        test_set[, 2:3], 
                                        min_vals, 
                                        max_vals))


# str

File: WORKSPACE/Assignment1/assignment1.py
#Author Aime Byiringiro

# We are  trying to predict if people in specific countries at certain age
# will or will not purchase 
# This is a classification problem since the dependent variable is categorical. 



#1. Importing libraries
import numpy as np
import pandas as pd
from sklearn.impute import SimpleImputer
from sklearn.compose import ColumnTransformer
from sklearn.preprocessing import OneHotEncoder
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import LabelEncoder
from sklearn.preprocessing import MinMaxScaler




#inputing dataset from Customer_Data.csv
dataset = pd.read_csv('Customer_Data.csv')


#Country, Age, and Salary will be independent(X)  and the last, Purchased, will
# be dependent y
X = dataset.iloc[:, :-1].to_numpy()
y = dataset.iloc[:, -1].to_numpy()


#Missing ages will be replacec by the mendian of the ages
imputer = SimpleImputer(missing_values=np.nan, strategy='median')
imputer.fit(X[:,1:2])
X[:,1:2] = imputer.transform(X[:,1:2])


# Missing salaries will be replaced the mean of salaries 
imputer = SimpleImputer(missing_values=np.nan, strategy='mean')
imputer.fit(X[:,2:3])
X[:,2:3] = imputer.transform(X[:,2:3])


# Since country is texture, we need to enncoder it to numerical
ct = ColumnTransformer(transformers = [('encoder', OneHotEncoder(), [0])], 
                       remainder='passthrough')
X = ct.fit_transform(X) 


#Encoding the last column of Purchased
le = LabelEncoder()
y = le.fit_transform(y)


# Feature scaling using normarization 
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=1/3, random_state=0)
sc =  MinMaxScaler()
X_train[:,3:] = sc.fit_transform(X_train[:,3:])
X_test[:,3:] = sc.transform(X_test[:,3:])















